specs/capri/OpenAPI_Design.md

```diff
@@ -0,0 +1,160 @@
+# VMware Ecosystem OpenAPI
+
+**Author(s)**: [zengyingzhe](https://github.com/zengyingzhe)
+This documentation serves as the design spec for OpenAPI.
+
+## Summary
+
+OpenAPI is a framework which provides generic storage related APIs to upper VMware plugins.
+
+## Motivation
+
+VMware is one of the most important and popular virtualization platform that depends on storage, so it is very necessory to integrate OpenSDS to VMware platform, and this will also benefit for OpenSDS' further development.
+To achieve various purposes, VMware supports multi types of plugins. All of them rely on storage APIs to interact with backend storage system.
+So if we implement a common framework that encapsulates API details of backend storage and provision more generic interfaces, it would be much more easier for plugin developers to implement a plugin, because developers can concern more to plugin itself's logic, not stuck in communication with storage.
+
+### Goals
+
+* Provide northbound generic APIs that encapsulate storage northbound APIs(through REST or others).
+* Open scalable&pluggable framework, 3rd parties can encapsulate its own storage's APIs following northbound interfaces' definition.
+
+### Non-Goals
+
+None.
+
+## Design Details
+
+### OpenAPI Usage Architecture
+
+![](media/OpenAPI_Usage_Architecture.png)
+OpenAPI is a language dependant library, such as a Jar package for Java.
+VMware plugins should import the OpenAPI library and call its northbound interfaces to communicate with corresponding storage.
+
+### OpenAPI Module
+
+![](media/OpenAPI_Module.png)
+OpenAPI provides unified northbound interfaces to VMware plugins, no matter what southbound storages connected.
+This could filter the differences between various of storages. Here we also see OpenSDS as a storage.
+To achieve the storage-not-aware purpose, each storage vendor must provide its own plugin to OpenAPI, which encapsulates storage calls based on storage's API definition.
+OpenAPI's northbound interfaces invokes these plugins' functions to actually interact with storages.
+
+### Data model impact
+
+No.
+
+### REST API impact
+
+May require OpenSDS to provide more REST APIs.
+
+### Security impact
+
+No.
+
+### Other end user impact
+
+No.
```

Comment:
Is there some impacts for client/osdsctl package changes? Please elaborate it if there is.